lessons/2_more_strings_and_functions/lesson.py

```diff
@@ -372,25 +419,85 @@ def inner_func():
+def sort_chars(string):
+    string_sorted = sorted(string)
+    string_sorted = "".join(string_sorted)
+    return string_sorted
+
+print(my_random_chars, "sorted is", sort_chars(my_random_chars))
+
+# # i realized this literally returns e.g. 'the total length of the string', 'l;k2;34lksd;lkfg', 'is', 16
+# # how do i have it return e.g. the total length of the string l;k2;34lksd;lkfg is 16
+
+# # # this seems to be why it is failing test.py
+# # # i think i fixed it, by just pulling the commentary out of the function and into the print.
+
+# 2. Write a function that takes a string and returns the value of the character at the index which is
+#    half of the total length of the string  if the string is even,
+#    the value of the penultimate character if the string is odd, 
 #    and 'Dobby' if an empty-string or length 1. Name this function dobby_search.
 
+def dobby_search(string):
+    if (len(string) % 2) == 0:
+        return string[len(string)//2]
+    else:
+        return string[len(string)-1]
+
+print(dobby_search(my_random_chars))
+
+# # # this one is failing the test, but i am not sure why.
+
 # 3. Write a function that takes in a string and return the ordinal value of the minimum char in the string
 #    or -1 if it is an empty string. Name the function min_to_ord.
 
+def min_to_ord(string):
+    if ord(min(string)) >= 0:
+        return ord(min(string))
+    else:
+        return -1
+
+print(min_to_ord(my_random_chars))
+
+
 # 4. Write a function takes in a string and returns the ordinal value of the maximum char in the string
 # or -1 if it is an empty string. Name the function max_to_ord.
 
-# 5. Write a function that takes in two argumnets: the first a string and the second a character. Return
+def max_to_ord(string):
+    if ord(max(string)) >= 0:
+        return ord(max(string))
+    else:
+        return -1
+
+print(max_to_ord(my_random_chars))
+
+# 5. Write a function that takes in two arguments: the first a string and the second a character. Return
 # how many times that character is found in the string. Name the function char_count.
 
+def char_count(string, char):
+    char_count = 0
+    for characters in string:
+        if characters == char:
+            char_count = char_count + 1
+        else:
+            char_count = char_count
+    return char_count
+
+print(char_count(my_random_chars, "2"))
+
+
 # 6. Imagine that you are scraping text data from the internet and want to print that data to your console. 
 #    Write a function that takes one argument, the data, and will return the data with all extra space stripped 
 #    away if any data is found and which returns "No data found!" if no data was found from scraping the internet. 
 #    Name the function pretty_format.
 
+def pretty_format(text):
+    if bool(text) == False:
+        return "No Data Found!"
+    else:
+        return text.strip()
```

Comment:
def pretty_format(string):
    if not string:
        return 'No data found!'
    return string.strip()